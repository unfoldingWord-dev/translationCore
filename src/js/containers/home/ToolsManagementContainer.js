import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
// components
import ToolsCards from "../../components/home/toolsManagement/ToolsCards";
import HomeContainerContentWrapper from "../../components/home/HomeContainerContentWrapper";
import {
  getTools,
  getIsUserLoggedIn,
  getProjectSaveLocation,
  getProjectBookId,
  getToolGatewayLanguage,
  getSourceBookManifest,
} from "../../selectors";
// actions
import { openTool } from "../../actions/ToolActions";
import { openAlertDialog } from "../../actions/AlertModalActions";
import * as PopoverActions from "../../actions/PopoverActions";
import * as ProjectDetailsActions from "../../actions/ProjectDetailsActions";
import { promptUserAboutMissingResource } from '../../actions/SourceContentUpdatesActions';
import * as BodyUIActions from '../../actions/BodyUIActions';
import {warnOnInvalidations} from "../../actions/ToolActions";

class ToolsManagementContainer extends Component {
  constructor(props) {
    super(props);
    this.handleSelectTool = this.handleSelectTool.bind(this);
  }

  shouldComponentUpdate(nextProps) {
    const {loadingProject} = nextProps.reducers.homeScreenReducer;
    if (loadingProject === true) {
      return false;
    } else return true;
  }

  componentDidMount() {
    const {tools, reducers} = this.props;
    const projectSaveLocation = getProjectSaveLocation(reducers);
    const bookId = getProjectBookId(reducers);
    if (projectSaveLocation && bookId) {
      tools.forEach(({name:toolName}) => {
        const currentGatewayLanguage = getToolGatewayLanguage(reducers, toolName);
        this.props.actions.loadCurrentCheckCategories(toolName, projectSaveLocation, currentGatewayLanguage);
      });
    }
  }

  handleSelectTool(toolName) {
    const {isUserLoggedIn, openTool, translate, openAlertDialog} = this.props;
    if(isUserLoggedIn) {
      openTool(toolName);
    } else {
      openAlertDialog(translate("please_log_in"));
    }
  }

  render() {
    const {
      tools,
      reducers: {
        loginReducer: { loggedInUser },
        projectDetailsReducer: {
          manifest,
          projectSaveLocation,
          toolsCategories
        },
        invalidatedReducer
      },
      translate,
      originalLanguageBookManifest,
      onMissingResource,
      toggleHomeView,
    } = this.props;

    const instructions = (
      <div>
        <p>
          {translate("tools.select_tool_from_list")}
        </p>
        <p>
          {translate("projects.supports_any_book", { app: translate("_.app_name") })}&nbsp;
          {translate(
            "projects.tools_required_resources",
            {
              actions: translate("actions"),
              check_for_content_updates: translate("updates.check_for_content_updates"),
            }
          )}
        </p>
      </div>
    );
    return (
      <HomeContainerContentWrapper
        translate={translate}
        instructions={instructions}
      >
        <div style={{ height: "100%" }}>
          {translate("tools.tools")}
          <ToolsCards
            tools={tools}
            onSelectTool={this.handleSelectTool}
            toolsCategories={toolsCategories}
            manifest={manifest}
            translate={translate}
            bookName={name}
            loggedInUser={loggedInUser}
            toggleHomeView={toggleHomeView}
            actions={{ ...this.props.actions }}
            onMissingResource={onMissingResource}
            originalLanguageBookManifest={originalLanguageBookManifest}
            invalidatedReducer={invalidatedReducer}
            projectSaveLocation={projectSaveLocation}
          />
        </div>
      </HomeContainerContentWrapper>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    isUserLoggedIn: getIsUserLoggedIn(state),
    tools: getTools(state),
    originalLanguageBookManifest: getSourceBookManifest(state),
    reducers: {
      homeScreenReducer: state.homeScreenReducer,
      projectDetailsReducer: state.projectDetailsReducer,
      loginReducer: state.loginReducer,
      invalidatedReducer: state.invalidatedReducer
    }
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    openTool: name => dispatch(openTool(name)),
    openAlertDialog: message => dispatch(openAlertDialog(message)),
    onMissingResource: (resourceDetails) => dispatch(promptUserAboutMissingResource(resourceDetails)),
    toggleHomeView: (value) => dispatch(BodyUIActions.toggleHomeView(value)),
    actions: {
      loadCurrentCheckCategories: (toolName, projectSaveLocation) => {
        dispatch(ProjectDetailsActions.loadCurrentCheckCategories(toolName, projectSaveLocation));
      },
      getProjectProgressForTools: (toolName, results) => {
        dispatch(ProjectDetailsActions.getProjectProgressForTools(toolName, results));
      },
      setProjectToolGL: (toolName, selectedGL) => {
        dispatch(ProjectDetailsActions.setProjectToolGL(toolName, selectedGL));
      },
      updateCategorySelection: (toolName, isChecked, subcategories) => {
        dispatch(ProjectDetailsActions.updateCategorySelection(toolName, isChecked, subcategories));
      },
      updateSubcategorySelection: (subcategory, toolName, isChecked) => {
        dispatch(ProjectDetailsActions.updateSubcategorySelection(subcategory, toolName, isChecked));
      },
      showPopover: (title, bodyText, positionCoord) => {
        dispatch(PopoverActions.showPopover(title, bodyText, positionCoord));
      },
      warnOnInvalidations: (toolName) => {
        dispatch(warnOnInvalidations(toolName));
      }
    }
  };
};

ToolsManagementContainer.propTypes = {
  isUserLoggedIn: PropTypes.bool.isRequired,
  openTool: PropTypes.func.isRequired,
  openAlertDialog: PropTypes.func.isRequired,
  tools: PropTypes.array.isRequired,
  reducers: PropTypes.shape({
    projectDetailsReducer: PropTypes.object.isRequired,
    loginReducer: PropTypes.shape({
      loggedInUser: PropTypes.bool
    }).isRequired,
    homeScreenReducer: PropTypes.object.isRequired
  }).isRequired,
  actions: PropTypes.object.isRequired,
  translate: PropTypes.func.isRequired,
  originalLanguageBookManifest: PropTypes.object.isRequired,
  onMissingResource: PropTypes.func.isRequired,
  toggleHomeView: PropTypes.func.isRequired,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ToolsManagementContainer);
