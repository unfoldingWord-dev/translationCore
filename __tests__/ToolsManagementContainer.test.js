/* eslint-env jest */
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import ToolsManagementContainer from '../src/js/containers/home/ToolsManagementContainer';
const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);


describe('ToolsManagementCotnainer container tests', () => {

  test('Check ToolsManagementContainer container instructions', () => {
    const props = {
      store: mockStore({
        toolsReducer: {
          toolsMetadata: []
        },
        loginReducer: {
          loggedInUser: true
        },
        settingsReducer: {
          currentSettings: {
            developerMode: false
          }
        },
        projectDetailsReducer: {
          manifest: {
            project: {
              id: ''
            }
          },
          projectSaveLocation: 'save/location',
          currentProjectToolsProgress: {},
          currentProjectToolsSelectedGL: {}
        }
      }),
      translate: k=>k
    };
    const wrapper = shallow(<ToolsManagementContainer {...props} />);
    expect(toJson(wrapper.dive())).toMatchSnapshot();
  });

});
