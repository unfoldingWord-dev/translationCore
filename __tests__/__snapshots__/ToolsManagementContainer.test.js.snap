// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToolsManagementCotnainer container tests Check ToolsManagementContainer container instructions 1`] = `
<HomeContainerContentWrapper
  instructions={
    <div>
      <p>
        tools.select_tool_from_list
      </p>
      <p>
        projects.books_available
      </p>
    </div>
  }
  translate={[Function]}
>
  <div
    style={
      Object {
        "height": "100%",
      }
    }
  >
    tools.tools
    <ToolsCards
      actions={
        Object {
          "getAllInvalidatedChecksForCurrentProject": [Function],
          "getProjectProgressForTools": [Function],
          "getToolsMetadatas": [Function],
          "launchTool": [Function],
          "setProjectToolGL": [Function],
        }
      }
      bookName="nodejs"
      currentProjectToolsProgress={Object {}}
      currentProjectToolsSelectedGL={Object {}}
      developerMode={false}
      loggedInUser={true}
      manifest={
        Object {
          "project": Object {
            "id": "",
          },
        }
      }
      projectSaveLocation="save/location"
      toolsMetadata={Array []}
      translate={[Function]}
    />
  </div>
</HomeContainerContentWrapper>
`;
